Error handling and propagation on My subscriptions refresh

anaspk opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes # .

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        | <img width="1071" alt="Screenshot 2025-03-26 at 3 57 39 PM" src="https://github.com/user-attachments/assets/18a7189f-84a2-4c6e-a1ae-151c7e6cc669" /> |
|        | <img width="1063" alt="Screenshot 2025-03-26 at 4 07 38 PM" src="https://github.com/user-attachments/assets/d5857d0f-023c-4372-adee-be966f69d8ca" /> |

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch on your local environment.
2. Build assets for WooCommerce Admin: `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build:admin`
3. Ensure that your WC Core development site is connected to your local WooCommerce.com development environment.
4. In the database of your WC Core development environment, in `wp_options` table,  corrupt the values of `access_token` and `access_token_secret` for the option_name `woocommerce_helper_data`.
5. Click the Refresh button on [My subscriptions page](http://wc-core.test:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions).
6. Notice that an error message is displayed indicating that the Authentication has failed.
7. Revert the changes you made to `woocommerce_helper_data` to make it corrupt in step 4 above.
8. On your local WooCommerce.com development environment, in `api-rate-limiting-check.php` , update the rate limit for /subscriptions endpoint to some low value (e.g. `3`) from the default value `120`.
9. Click the Refresh button on [My subscriptions page](http://wc-core.test:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions).
10. Notice that an error message is displayed indicating that the WooCommerce.com API returned HTTP status code 429.
11. While the subscriptions refresh is failing due to either of the above errors, browse different areas of the site to make sure that nothing else is broken due to the changes in this PR e.g. none of the exceptions are left unhandled in any of the flows.
12. Also try refreshing the [My subscriptions page](http://wc-core.test:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) completely (by clicking the Refresh button on the browser) to make sure that it still works as expected.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

Handle and display any errors that occur while refreshing WooCommerce.com subscriptions list

</details>

anaspk committed:
Error propagation on My subscriptions refresh - first draft

anaspk committed:
Remove an unnecessary try/catch

anaspk committed:
Add default exception handling to all unrelated usages

anaspk committed:
Add exception handling to a place missed previously

anaspk committed:
Fix wrong property name for error message in response

anaspk committed:
Improve exception error messages

anaspk committed:
Merge branch 'trunk' into fix/subscriptions-refresh-error-propagation

anaspk committed:
Fix linting issues

anaspk committed:
Merge branch 'fix/subscriptions-refresh-error-propagation' of github.com:woocommerce/woocommerce into fix/subscriptions-refresh-error-propagation

Commit:
Add changefile(s) from automation for the following project(s): woocommerce, woocommerce/client/admin

anaspk committed:
Fix a couple more linting issues

anaspk committed:
Merge remote-tracking branch 'origin/fix/subscriptions-refresh-error-propagation' into fix/subscriptions-refresh-error-propagation

anaspk committed:
Merge branch 'trunk' into fix/subscriptions-refresh-error-propagation

thilinah commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1691,43 +1717,53 @@ public static function get_subscriptions() {
-		}
+			$code = wp_remote_retrieve_response_code( $request );
+			if ( 200 !== $code ) {
+				// translators: %d: HTTP status code.
```
In case of a non 200 response, we need to set the transient to an empty array. Like in the previous implementation. ( E.g `set_transient( $cache_key, array(), 15 * MINUTE_IN_SECONDS );`).

This logic was there to prevent a store from keep calling WCCOM in short intervals when subscription request is failing, for an example due to rate-limiter.

anaspk replied:
Added it back in https://github.com/woocommerce/woocommerce/pull/56613/commits/c98b6ed5a368362dbd0a9b3929417bf9c0abcbab

thilinah commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1589,27 +1605,36 @@ public static function get_product_usage_notice_rules() {
-			$data = array();
-		}
+			$code = wp_remote_retrieve_response_code( $request );
+			if ( 200 !== $code ) {
```
Same issue is also here: https://github.com/woocommerce/woocommerce/pull/56613/files#r2011573478

anaspk replied:
Added it back in https://github.com/woocommerce/woocommerce/pull/56613/commits/c98b6ed5a368362dbd0a9b3929417bf9c0abcbab

thilinah commented on plugins/woocommerce/includes/admin/helper/class-wc-helper.php:
```diff
@@ -1691,43 +1717,53 @@ public static function get_subscriptions() {
+			$code = wp_remote_retrieve_response_code( $request );
+			if ( 200 !== $code ) {
+				// translators: %d: HTTP status code.
+				throw new Exception( sprintf( __( 'WooCommerce.com API returned HTTP status code %d.', 'woocommerce' ), $code ) );
```
The most common error states for subscription request is 429 and 403. For these response codes, rather than showing a generic status code, based error, we can make it more descriptive.

E.g: 
- for 429: You have exceeded the request limit. Please try again after a few minutes.
- for 403: We're having trouble authenticating your store. Please try again. If the issue persists, disconnect your store from WooCommerce.com and reconnect.

hesahesa replied:
Agree with @thilinah that we probably should make the message more descriptive and not show the HTTP status code in the notice?

anaspk replied:
Thanks for the suggestions @thilinah , I have updated the error messages in https://github.com/woocommerce/woocommerce/pull/56613/commits/c98b6ed5a368362dbd0a9b3929417bf9c0abcbab and https://github.com/woocommerce/woocommerce/pull/56613/commits/7704452854db9354562328c7145b50b04f4c8e4f 

@hesahesa I think it is fine to show the HTTP status code in case of less common errors as it will help with debugging both the user themselves and also our HE if they contact support. WDYT?

hesahesa replied:
@anaspk yes i think on a less common HTTP status code we can show it in the notice

anaspk committed:
Improvements around caching and error messages in case of error

anaspk committed:
Update authentication failure error message on another place

anaspk commented:
Thanks for the review @thilinah , I have addressed your feedback and also updated the screenshots in issue description with the updated error messages. Please take another look when you have time.

anaspk committed:
Merge branch 'trunk' into fix/subscriptions-refresh-error-propagation

hesahesa commented on plugins/woocommerce/client/admin/client/marketplace/components/my-subscriptions/table/actions/refresh-button.tsx:
```diff
@@ -44,7 +44,7 @@ export function RefreshButton() {
 							'woocommerce'
 						),
-						error.message
+						error.data.message
```
i'm trying to understand why we need the `error.data` here?

the `error` is an instance of this, right? https://developer.mozilla.org/en-US/docs/Web/API/Window/fetch#exceptions or is it a different thing?

anaspk replied:
@hesahesa I think the structure of response inside this error object is whatever is returned from backend. In our case, we return response using [`wp_send_json_error` here](https://github.com/woocommerce/woocommerce/pull/56613/files#diff-4cf137656b8d29bec4423f31fb2781a2d67d2866f172c2620ae3a5d4a2d75294R152). `wp_send_json_error` [automatically wraps](https://developer.wordpress.org/reference/functions/wp_send_json_error/#more-information) whatever `value` we provide it in a top level `data` key. That's what we are unwrapping here.

anaspk committed:
Remove redundant retry button from error notice in RefreshButton component.

anaspk committed:
Add doc comment for a function that was missing it

anaspk committed:
Merge branch 'trunk' into fix/subscriptions-refresh-error-propagation

anaspk commented:
@thilinah The space was coming from the flex gap applied to the whole notice component via the CSS.

<img width="1057" alt="Screenshot 2025-03-28 at 11 58 27 AM" src="https://github.com/user-attachments/assets/5951332a-6a90-4e7a-bb05-cb4cc919ab81" />

I have removed the Try again button in https://github.com/woocommerce/woocommerce/pull/56613/commits/dfceadb67c2e32f9e6bc53fbb7a686781b861d8e . Now it looks like this:

<img width="1045" alt="Screenshot 2025-03-28 at 12 00 58 PM" src="https://github.com/user-attachments/assets/e940e6c3-b2ca-4e38-85df-be8907552013" />

thilinah reviewed (approved):
Thanks @anaspk, LGTM, tested and works as expected.
